File: employee_ends_service_award.py
# -*- coding: utf-8 -*-
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt
from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.model.document import Document
import json
import math
from frappe.utils import cint, cstr, date_diff, flt, formatdate, getdate, get_link_to_form, \
    comma_or, get_fullname, add_years, add_months, add_days, nowdate,get_first_day,get_last_day
import datetime
from datetime import date
from frappe.utils import date_diff
from datetime import datetime
from dateutil import relativedelta


class EmployeeEndsServiceAward(Document):

    def on_submit(self):
        emp = frappe.get_doc("Employee", self.employee)
        emp.status = "Left"
        emp.relieving_date = getdate(get_last_day(add_months(getdate(self.end_date),-1)))
        emp.save(ignore_permissions=True)
        ss = frappe.get_doc("Salary Structure", {"employee":self.employee}, "name")
        ss.is_active = "No"
        ss.save(ignore_permissions=True)
        

    def get_days_months_years(self):
        difference = relativedelta.relativedelta(getdate(self.end_date), getdate(self.work_start_date))
        return difference.years, difference.months, difference.days + 1


        
        
    def get_salary(self,employee):
        basic_salary = 0
        housing_allowance = 0
        transportation_allowance = 0
        salary_structure = frappe.get_value("Salary Structure", {"employee": self.employee}, "name")
        if salary_structure:
            result = frappe.db.sql("select sum(amount) from `tabSalary Detail` where parent='{0}' and parentfield = 'earnings' and salary_component in ('Basic','Housing Allowance','Transportation Allowance', 'Deduction Comp', 'بدل انتقال', 'Subsistence allowance', 'Social Security','Management allowance', 'Telecom allowance', 'Other allowances', 'Internet allowance')".format(salary_structure))
            #result = frappe.db.sql("select sum(amount) from `tabSalary Detail` where parent='{0}' and parentfield = 'earnings' and salary_component in ('اساسي','بدل اتصال','بدل اعاشة', 'بدل مواصلات 'بدل بدلات اخرى', 'بدل انترنت','بدل ادارة','بدل سكن', 'عمل اضافي', 'اتعاب ادارية', 'التامينات الاجتماعية', 'اخرى', 'جزاءات','مكافئة', 'سلف)".format(salary_structure))
            if result:
                result2 = frappe.db.sql("select sum(amount) from `tabSalary Detail` where parent='{0}' and parentfield = 'earnings' and salary_component in ('Basic','Housing Allowance')".format(salary_structure))
                basic_salary = frappe.get_value("Salary Detail", {"parent": salary_structure, "parentfield": 'earnings', "salary_component": 'Basic'}, "amount") or 0
                housing_allowance = frappe.get_value("Salary Detail", {"parent": salary_structure, "parentfield": 'earnings', "salary_component": 'Housing Allowance'}, "amount") or 0
                transportation_allowance = frappe.get_value("Salary Detail", {"parent": salary_structure, "parentfield": 'earnings', "salary_component": 'Transportation Allowance'}, "amount") or 0
                
                return cstr(result[0][0]), round(result[0][0]/30, 2), round(result2[0][0]/30, 2), basic_salary, housing_allowance, transportation_allowance
            else:
                frappe.throw(_("No salary structure found for this employee"))
        else:
            frappe.throw(_("No salary structure found for this employee"))


    def get_leave_balance(self,employee):
        total_leave_balance = frappe.db.sql("select total_leaves_allocated,from_date,to_date,name from `tabLeave Allocation` where employee='{0}' and leave_type='Annual Leave - اجازة اعتيادية' order by creation desc limit 1".format(employee))
        if total_leave_balance:
            leave_days = frappe.db.sql("select sum(total_leave_days) from `tabLeave Application` where employee='{0}' and leave_type='Annual Leave - اجازة اعتيادية' and posting_date between '{1}' and '{2}'".format(employee,total_leave_balance[0][1],total_leave_balance[0][2]))[0][0]
            if not leave_days:
                leave_days = 0
            leave_balance =  int(total_leave_balance[0][0])-int(leave_days)
        else:
            leave_balance = 0
        return leave_balance
        
    
@frappe.whitelist()
def get_award(start_date, end_date, salary, toc, reason):
    start = start_date
    end = end_date
    ret_dict = {}

    if getdate(end) < getdate(start):
        frappe.throw("تاريخ نهاية العمل يجب أن يكون أكبر من تاريخ بداية العمل")
    else:
        diffDays = date_diff(end, start)
        years = math.floor(diffDays / 360)
        daysrem = diffDays - (years * 360)
        months = math.floor(daysrem / 30)
        days = math.ceil(daysrem - (months * 30))
        ret_dict = {"days":days, "months":months, "years":years, "award":0}
    years = flt(years) + (flt(months) / 12) + (flt(days) / 360)
    if not reason:
        frappe.throw("برجاء اختيار سبب انتهاء العلاقة العمالية")
    else:
        if  "كفالة فقط" in toc:
            if reason == "":
                ret_dict["award"] = 0
            else:
                firstPeriod = 0
                secondPeriod = 0
                if years > 5:
                    firstPeriod = 5
                    secondPeriod = years - 5
                else:
                    firstPeriod = years
                result = (firstPeriod * salary * 0.5) + (secondPeriod * salary)
                ret_dict["award"] = result
        else:

            if reason == "فسخ العقد":
                ret_dict["award"] = 0
            elif reason =="استقالة الموظف قبل انتهاء مدة العقد":
                if years < 2:
                    result = 0
                elif years <= 5:
                    result = (1.0 / 6.0) *  salary * years 
                elif years <= 10:
                    result = ((1.0 / 3.0) *  salary * 5) + ((2.0 / 3.0) *  salary * (years - 5))
                else:
                    result = (0.5 *  salary * 5) + ( salary * (years - 5))
                ret_dict["award"] = result
            else:
                if years <= 5:
                    result = 0.5 *  salary * years
                else:
                    result = (0.5 *  salary * 5) + salary * (years - 5)
                ret_dict["award"] = result
               
    return ret_dict
        
        

File: employee_ends_service_award.js
// Copyright (c) 2022, Ahmed and contributors
// For license information, please see license.txt

cur_frm.add_fetch("employee", "date_of_joining", "work_start_date");
frappe.ui.form.on('Employee Ends Service Award',{
    setup: function(frm) {

        frm.set_query("employee", function() {
            return {
                filters: {
                    status: 'Active',
                }
            };
        });
    },
    refresh: function(frm) {

        if (cur_frm.doc.employee) {

            if (cur_frm.doc.type_of_contract == "عمل عن بعد" || cur_frm.doc.type_of_contract == "عقد وظيفي جزئي" || cur_frm.doc.type_of_contract =="عقود موظفين" || cur_frm.doc.type_of_contract =="عقود عمال" || cur_frm.doc.type_of_contract == "عقد أجير" || cur_frm.doc.type_of_contract == "كفالة فقط") {
				cur_frm.set_df_property("reason", "options", "\nانتهاء مدة العقد أو الاتفاق بين الطرفين على انهاء العقد أو انهاء العقد من قبل الشركة\nاستقالة الموظف قبل انتهاء مدة العقد");
            } else {
                cur_frm.set_df_property("reason", "options", "");
            }
        }
    },
    notice_month: function(frm) {
        if (frm.doc.notice_month) {
            frm.set_df_property("notice_month_start", "reqd", true);
            frm.set_df_property("notice_month_end", "reqd", true);
        } else {
            frm.set_df_property("notice_month_start", "reqd", false);
            frm.set_df_property("notice_month_end", "reqd", false);
        }
    },
    employee: function(frm) {

        if (cur_frm.doc.employee) {
            frappe.call({
                "method": "get_salary",
                doc: cur_frm.doc,
                args: { "employee": cur_frm.doc.employee },
                callback: function(data) {
                    if (data) {
                        cur_frm.set_value('salary', data.message[0]);
                        cur_frm.set_value('day_value', data.message[1]);
                        cur_frm.set_value('leave_cost', data.message[2]);

                        cur_frm.set_value('basic', data.message[3]);
                        cur_frm.set_value('housing_allowance', data.message[4]);
                        cur_frm.set_value('transportation_allowance', data.message[5]);
                    }
                }
            });
        }

        if (cur_frm.doc.employee && cur_frm.doc.end_date) {
            var end_resignation_date = new Date(cur_frm.doc.end_date);

            var date1 = new Date(cur_frm.doc.work_start_date);
            var date2 = new Date(cur_frm.doc.end_date);
            var diffTime = date2.getTime() - date1.getTime();
            //~ var diffTime = Math.abs(date2 - date1);
            //~ var diffDays = diffTime / (1000 * 3600 * 24);
            var diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
            
            
            

            if (diffDays < 30) {
                cur_frm.set_value("days_number", diffDays)
            } else {
                cur_frm.set_value("days_number", end_resignation_date.getDate())
            }

            cur_frm.set_value("total_month_salary", cur_frm.doc.days_number * cur_frm.doc.day_value)
        }

    },
    salary: function(frm) {
        if (cur_frm.doc.salary) {
            frappe.call({
                "method": "get_leave_balance",
                doc: cur_frm.doc,
                args: { "employee": cur_frm.doc.employee },
                callback: function(data) {
                    if (data) {
                        cur_frm.set_value('leave_number', data.message);
                        cur_frm.set_value('leave_total_cost', Math.round(data.message * cur_frm.doc.leave_cost));
                        
                        
                    }
                }
            });
        }
    },
    
    
      type_of_contract: function(frm) {
        if (cur_frm.doc.type_of_contract == "عمل عن بعد" || cur_frm.doc.type_of_contract == "عقد وظيفي جزئي" || cur_frm.doc.type_of_contract =="عقود موظفين" || cur_frm.doc.type_of_contract =="عقود عمال") {
			cur_frm.set_df_property("reason", "options", "\nانتهاء مدة العقد أو الاتفاق بين الطرفين على انهاء العقد أو انهاء العقد من قبل الشركة\nاستقالة الموظف قبل انتهاء مدة العقد");
        } else {
			cur_frm.set_df_property("reason", "options", "");
        }
    },
   end_date: function(frm) {


        frm.trigger("get_days_months_years");


        if (cur_frm.doc.employee && cur_frm.doc.end_date) {

            var end_resignation_date = new Date(cur_frm.doc.end_date);


            var date1 = new Date(cur_frm.doc.work_start_date);
            var date2 = new Date(cur_frm.doc.end_date);
            //~ var diffTime = date2.getTime() - date1.getTime();
            //~ var diffDays = diffTime / (1000 * 3600 * 24);
            var diffTime = Math.abs(date2 - date1);
            var diffDays = Math.ceil(diffTime / (1000 * 3600 * 24 ));
            
            

            if (diffDays < 30) {
                cur_frm.set_value("days_number", diffDays)
            } else {
                cur_frm.set_value("days_number", end_resignation_date.getDate())
            }

            cur_frm.set_value("total_month_salary", cur_frm.doc.days_number * cur_frm.doc.day_value)
        }

    },
    //~ get_days_months_years: function(frm) {
        //~ var start = cur_frm.doc.work_start_date;
        //~ var end = cur_frm.doc.end_date;

        //~ if (end < start) {
            //~ cur_frm.set_value('years', 0);
            //~ cur_frm.set_value('months', 0);
            //~ cur_frm.set_value('days', 0);
            //~ validated = false;
            //~ frappe.throw("تاريخ نهاية العمل يجب أن يكون أكبر من تاريخ بداية العمل");
            

        //~ } else {
            //~ var date1 = new Date(start);
            //~ var date2 = new Date(end);
            //~ var oneDay = 1000*60*60*24;
            //~ var timeDiff = Math.abs(date2.getTime() - date1.getTime());
            //~ var diffDays = Math.ceil(timeDiff /oneDay);
            //~ var years = Math.floor(diffDays / 365);
            //~ var daysrem = diffDays - (years * 365);
            //~ var months = Math.floor(daysrem / 30);
            //~ var monthss = months
            //~ var days = Math.ceil(daysrem - (months * 30));

            //~ cur_frm.set_value('years', years);
            //~ cur_frm.set_value('months', monthss);
            //~ cur_frm.set_value('days', days);

        //~ };


    //~ },
	get_days_months_years: function(frm) {
        var start = cur_frm.doc.work_start_date;
        var end = cur_frm.doc.end_date;

        if (end < start) {
            cur_frm.set_value('years', 0);
            cur_frm.set_value('months', 0);
            cur_frm.set_value('days', 0);
            validated = false;
            frappe.throw("تاريخ نهاية العمل يجب أن يكون أكبر من تاريخ بداية العمل");


        } else {


            frappe.call({
                "method": "get_days_months_years",
                doc: cur_frm.doc,
                callback: function(data) {
                    if (data) {
						cur_frm.set_value('years', data.message[0]);
						cur_frm.set_value('months', data.message[1]);
						cur_frm.set_value('days', data.message[2]);
					} else {
						cur_frm.set_value('years', 0);
						cur_frm.set_value('months', 0);
						cur_frm.set_value('days', 0);
					}
                }
            });


        };


    },
 
    validate: function(frm) {

        var total = 0;
        $.each(frm.doc.end_of_service_award_deduction || [], function(i, d) {
            total += parseFloat(d.deduction);
        });

        cur_frm.set_value("total_deduction", total);



        var total = 0;
        $.each(frm.doc.end_of_service_award_earning || [], function(i, d) {
            total += parseFloat(d.earning);
        });

        cur_frm.set_value("total_earning", total);


        frm.trigger("get_award");
    },
    after_save: function(frm) {
        var notice_month_salary = 0
        if (cur_frm.doc.notice_month) {
            notice_month_salary = parseFloat(cur_frm.doc.salary)
        }

        cur_frm.set_value('total', cur_frm.doc.award + cur_frm.doc.total_earning - cur_frm.doc.total_deduction);
    },
    find: function(frm) {
        frm.trigger("get_award");
    },
    ticket_number: function(frm) {
        if (cur_frm.doc.ticket_number && cur_frm.doc.ticket_cost) {
            cur_frm.set_value('ticket_total_cost', cur_frm.doc.ticket_number * cur_frm.doc.ticket_cost);
        }
    },
    ticket_cost: function(frm) {
        if (cur_frm.doc.ticket_number && cur_frm.doc.ticket_cost) {
            cur_frm.set_value('ticket_total_cost', cur_frm.doc.ticket_number * cur_frm.doc.ticket_cost);
        }
    },

    get_award: function(frm) {
        var result = 0
        if (!cur_frm.doc.reason) {
            frappe.throw("أرجو اختيار سبب نهاية الخدمة");
        }
        cur_frm.set_value('award', 0);

        var salary = cur_frm.doc.salary;
        var years = parseInt(cur_frm.doc.years) + (parseInt(cur_frm.doc.months) / 12) + (parseInt(cur_frm.doc.days) / 360);
        var reason = cur_frm.doc.reason;

        if (!reason) {
            frappe.throw("برجاء اختيار سبب انتهاء العلاقة العمالية");
            cur_frm.set_value('award', 0);
        } else { 
			if (cur_frm.doc.reason == "انتهاء مدة العقد أو الاتفاق بين الطرفين على انهاء العقد أو انهاء العقد من قبل الشركة"){
                    // cur_frm.set_value('award', "");
                    var firstPeriod, secondPeriod = 0;
                    // set periods
                    if (years > 5) {
                        firstPeriod = 5;
                        secondPeriod = years - 5;
                    } else {
                        firstPeriod = years;
                    }
                    // calculate
                    result = (firstPeriod * cur_frm.doc.salary * 0.5) + (secondPeriod * cur_frm.doc.salary);
                    cur_frm.set_value('award', Math.round(result*100)/100);
			} else {
                if (cur_frm.doc.reason == "") {
                    cur_frm.set_value('award', 0);
                } else if (cur_frm.doc.reason =="استقالة الموظف قبل انتهاء مدة العقد") {
                    if (years < 2) {
                        result = 0;
                    } else if (years <= 5) {
                        result = (1 / 6) * cur_frm.doc.salary * years;
                    } else if (years <= 10) {
                        result = ((1 / 3) * cur_frm.doc.salary * 5) + ((2 / 3) * cur_frm.doc.salary * (years - 5));
                    } else {
                        result = (0.5 * cur_frm.doc.salary * 5) + (cur_frm.doc.salary * (years - 5));
                    }
                    if (typeof(result) === 'number') {
                        cur_frm.set_value('award', Math.round(result*100)/100 );
                    } else {
                        cur_frm.set_value('award', Math.round(result*100)/100);
                    }
                } else {
                    if (years <= 5) {
                        result = 0.5 * cur_frm.doc.salary * years;
                    } else {
                        result = (0.5 * cur_frm.doc.salary * 5) + (cur_frm.doc.salary * (years - 5));
                    }
                    if (typeof(result) === 'number') {
                        cur_frm.set_value('award', Math.round(result*100)/100);
                    } else {
                        cur_frm.set_value('award', Math.round(result*100)/100);
                    }
                    console.log(Math.round(result*100)/100);
                }


            }
        };
    }

});






frappe.ui.form.on('End of Service Award Deduction', {
    deduction: function(frm, cdt, cdn) {
        var row = locals[cdt][cdn];

        var total = 0;
        $.each(frm.doc.end_of_service_award_deduction || [], function(i, d) {
            total += parseFloat(d.deduction);
        });

        cur_frm.set_value("total_deduction", total);

    }

});





frappe.ui.form.on('End of Service Award Earning', {
    earning: function(frm, cdt, cdn) {
        var row = locals[cdt][cdn];

        var total = 0;
        $.each(frm.doc.end_of_service_award_earning || [], function(i, d) {
            total += parseFloat(d.earning);
        });

        cur_frm.set_value("total_earning", total);

    }

});
